File: data/gamification.js

export const BADGES = [
  {
    id: 'first-step',
    title: 'First Step',
    emoji: '🥉',
    desc: 'First topic complete! Welcome to the journey.',
    requirement: { type: 'count', value: 1 }
  },
  {
    id: 'explorer',
    title: 'Explorer',
    emoji: '🥈',
    desc: '10 topics completed! You are getting deep into tech.',
    requirement: { type: 'count', value: 10 }
  },
  {
    id: 'tech-learner',
    title: 'Tech Learner',
    emoji: '🥇',
    desc: '25 topics completed! Consistency is your middle name.',
    requirement: { type: 'count', value: 25 }
  },
  {
    id: 'tech-master',
    title: 'Tech Master',
    emoji: '💎',
    desc: '50 topics completed! You are now a knowledge beast.',
    requirement: { type: 'count', value: 50 }
  },
  {
    id: 'centurion',
    title: 'Centurion',
    emoji: '💯',
    desc: '100 topics completed! Legend level reached.',
    requirement: { type: 'count', value: 100 }
  },
  {
    id: 'web-basics',
    title: 'Web Basics',
    emoji: '🌐',
    desc: 'All Level 1 & 2 basics topics completed.',
    requirement: { type: 'levels', value: [1, 2] }
  },
  {
    id: 'deployment-master',
    title: 'Deployment Master',
    emoji: '🚀',
    desc: 'All Level 4 hosting topics completed.',
    requirement: { type: 'levels', value: [4] }
  },
  {
    id: 'ai-beginner',
    title: 'AI Beginner',
    emoji: '🤖',
    desc: 'All Level 7 AI topics completed.',
    requirement: { type: 'levels', value: [7] }
  },
  {
    id: 'git-master',
    title: 'Git Master',
    emoji: '🧑‍💻',
    desc: 'Specific Git tools completed.',
    requirement: { type: 'ids', value: ['git-basics', 'github-basics', 'git-commit', 'git-branch'] }
  },
  {
    id: 'night-owl',
    title: 'Night Owl',
    emoji: '🦉',
    desc: 'Learn after 12 AM.',
    requirement: { type: 'time', value: '00-05' }
  }
];

export const XP_LEVELS = [
  { level: 1, name: 'Curious Beginner', minXP: 0 },
  { level: 2, name: 'Tech Explorer', minXP: 200 },
  { level: 3, name: 'Developer', minXP: 500 },
  { level: 4, name: 'Tech Pro', minXP: 1000 },
  { level: 5, name: 'Tech Architect', minXP: 2500 },
  { level: 6, name: 'CTO Level', minXP: 5000 },
  { level: 7, name: 'Tech Legend', minXP: 10000 },
];

export const XP_AWARDS = {
  READ_TOPIC: 10,
  QUIZ_CORRECT: 20,
  DAILY_LOGIN: 5,
  SHARE_TOPIC: 15
};

File: techaa-backend/internal/gamification/gamification.go
package gamification

import "math"

type XPLevel struct {
	Level int    `json:"level"`
	Title string `json:"title"`
	MinXP int    `json:"min_xp"`
	MaxXP int    `json:"max_xp"`
}

type BadgeDef struct {
	ID          string `json:"id"`
	Title       string `json:"title"`
	Description string `json:"description"`
	IconEmoji   string `json:"icon_emoji"`
	Requirement string `json:"requirement"`
	XPReward    int    `json:"xp_reward"`
}

var (
	XPAwards = map[string]int{
		"READ_TOPIC":       10,
		"QUIZ_CORRECT":     20,
		"DAILY_LOGIN":      5,
		"SHARE_TOPIC":      15,
		"COURSE_COMPLETED": 100,
	}

	XPLevels = []XPLevel{
		{Level: 1, Title: "Curious Beginner", MinXP: 0, MaxXP: 199},
		{Level: 2, Title: "Tech Explorer", MinXP: 200, MaxXP: 499},
		{Level: 3, Title: "Developer", MinXP: 500, MaxXP: 999},
		{Level: 4, Title: "Tech Pro", MinXP: 1000, MaxXP: 2499},
		{Level: 5, Title: "Tech Architect", MinXP: 2500, MaxXP: 4999},
		{Level: 6, Title: "CTO Level", MinXP: 5000, MaxXP: 9999},
		{Level: 7, Title: "Tech Legend", MinXP: 10000, MaxXP: 999999},
	}

	Badges = []BadgeDef{
		{
			ID:          "badge_first_step",
			Title:       "First Step",
			Description: "Completed your very first tech topic!",
			IconEmoji:   "🚀",
			Requirement: "Complete 1 topic",
			XPReward:    20,
		},
		{
			ID:          "badge_quiz_warrior",
			Title:       "Quiz Warrior",
			Description: "Scored 100% on a battle quiz.",
			IconEmoji:   "⚔️",
			Requirement: "Score 100% in a quiz",
			XPReward:    30,
		},
		{
			ID:          "badge_streak_master",
			Title:       "Streak Master",
			Description: "Maintained a 7-day learning streak!",
			IconEmoji:   "🔥",
			Requirement: "Reach 7 days streak",
			XPReward:    50,
		},
		{
			ID:          "badge_web_explorer",
			Title:       "Web Explorer",
			Description: "Completed foundational Web topics.",
			IconEmoji:   "🌐",
			Requirement: "Complete Web topics",
			XPReward:    40,
		},
		{
			ID:          "badge_cloud_ninja",
			Title:       "Cloud Ninja",
			Description: "Mastered Docker, DevOps, and Cloud deployment.",
			IconEmoji:   "☁️",
			Requirement: "Complete Cloud course",
			XPReward:    60,
		},
		{
			ID:          "badge_ai_pioneer",
			Title:       "AI Pioneer",
			Description: "Explored Large Language Models and prompt engineering.",
			IconEmoji:   "🤖",
			Requirement: "Complete AI topics",
			XPReward:    80,
		},
	}
)

type LevelInfo struct {
	Level          int     `json:"level"`
	Title          string  `json:"title"`
	TotalXP        int     `json:"total_xp"`
	MinXP          int     `json:"min_xp"`
	MaxXP          int     `json:"max_xp"`
	LevelProgress  float64 `json:"level_progress"`
	XPToNextLevel  int     `json:"xp_to_next_level"`
}

func CalculateLevel(totalXP int) LevelInfo {
	cur := XPLevels[0]
	var next *XPLevel

	for i := range XPLevels {
		if totalXP >= XPLevels[i].MinXP {
			cur = XPLevels[i]
			if i+1 < len(XPLevels) {
				next = &XPLevels[i+1]
			} else {
				next = nil
			}
		}
	}

	progress := 1.0
	xpToNext := 0

	if next != nil {
		xpInLevel := totalXP - cur.MinXP
		xpSpan := next.MinXP - cur.MinXP
		if xpSpan > 0 {
			progress = math.Min(1.0, math.Max(0.0, float64(xpInLevel)/float64(xpSpan)))
		}
		xpToNext = int(math.Max(0, float64(next.MinXP-totalXP)))
	}

	return LevelInfo{
		Level:         cur.Level,
		Title:         cur.Title,
		TotalXP:       totalXP,
		MinXP:         cur.MinXP,
		MaxXP:         cur.MaxXP,
		LevelProgress: math.Round(progress*1000) / 1000,
		XPToNextLevel: xpToNext,
	}
}
